_posts/2020-05-08-coupling-and-cohesion.markdown

```diff
@@ -0,0 +1,138 @@
+*cohesion* to guide your design and refactoring decisions yielded
+immediate results for me.
+
+<!--more-->
+
+Judging from a lot of the code that I've had
+to work on over the years, it could help quite a few other developers
+as well.
+
+## The problem in a nutshell
+
+A lot of my early design efforts tended to wind up with amorphously
+defined "layers," catch-all classes full of random junk, and clunky
+abstractions. I knew about a laundry list of design principles like
+SOLID and DRY, [abstraction](https://simpleprogrammer.com/respecting-abstraction)
+and design patterns. I knew that classes should "talk" to each other
+as little as possible. But, when it came time to actually write (or
+improve) code, the results were always somehow less than satisfactory.
+Ideally, we want neatly factored designs with relatively small methods
+and classes, each of which only does one thing, with zero side
+effects. But as I tried to reach that ideal, I seemed to always wind
+up with pieces - sometimes large pieces - of code that didn't
+obviously belong in any particular place. Worse, they couldn't be put
+anywhere without bloating already existing classes or otherwise
+ruining the organization of the code. The question of "where should
+this code live" is something that every working developer will run
+into over and over again.
+
+Learning about the twin concepts of coupling and cohesion suddenly
+clarified why I had been having trouble. Even better, all of the
+other design principles suddenly made a much deeper and more cohesive
+sense - all of them were just different ways of addressing these two
+key principles.
+
+## Coupling
+
+Of the two concepts,
+[coupling](https://en.wikipedia.org/wiki/Coupling_%28computer_programming%29)
+is the one that most programmers are more likely to be familiar with.
+Coupling refers to the degree to which two components - classes or
+modules - interact with one another. It's a measure of how much they
+"know" about each other. When you're reminded to keep methods that
+aren't called by other classes private, you're hearing a half-hearted
+stab at coupling. Marking methods private is just one way to reduce
+the number of methods that any other part of the code can see or know
+about. But coupling can also occur through data - which is why using
+a global variable, for instance, is a code smell. Any class or module
+that references the global variable is now coupled to all of the
+others through that variable. Generally speaking, we want to reduce
+coupling to the bare minimum needed to implement a solution.
+Conveniently, there is a hierarchy of types of coupling. Personally,
+I've found that learning about this hierarchy, and at least keeping it
+in mind when designing software is worth the effort.
+
+
+## Cohesion
+
+The twin idea of Cohesion, for whatever reason, seems to be discussed
+less often, probably because it can be a bit more abstract at first.
+It's really just the inverse concept of coupling.
```

Comment:
I don't get this sentence: How is it the inverse concept of coupling? 

I understand that they're interconnected... Usually when you increase cohesion for a module (A), you are extracting a new module (B), having them "talk" to each other, so you're increasing cohesion per module, but also increasing coupling (from no coupling to some coupling)